### PR TITLE
Fix: emissive hack

### PR DESCRIPTION
**Pull request by s520**
If you combine textures with emission colors, the new renderer will make the colors brighter.

- Old renderer
Emissive_OR

- New renderer (Emissive branch #525)
Emissive_NR

- New renderer (This PR)
Emissive_NR_Fix

I also removed and cleaned up unused variables in the shader.

### CODE REVIEW

**Comment by leezer3** [comment]
Thanks!

Hopefully we're slowly getting somewhere, and might get this properly right at some stage.....
I'll try testing this properly later.

Also requires this reverting:
https://github.com/leezer3/OpenBVE/commit/82e43116648c53933d8178c0e0a9f9fa793c8048

---

I think at this point, we probably need a list of known issues / edge cases that have been fixed in the shader, otherwise things get confused and broken again.

File: assets/Shaders/default.frag [hunk 1]
```diff
@@ -15,36 +16,31 @@ uniform float uFogEnd;
...
-uniform int uMaterialFlags;
 
 void main(void)
 {
-	vec4 textureColor = vec4(oColor);
-	if(uIsTexture)
-	{
-		textureColor *= texture2D(uTexture, oUv);
-	}
+	vec4 finalColor = uIsTexture ? texture2D(uTexture, oUv) : vec4(oColor.rgb, 1.0);
```
Comment (leezer3) [inline assets/Shaders/default.frag:22]: Minor:
You're assuming colored vertices disable texturing.

At the minute, I don't think there's anything with colored vertices in the wild (only format that supports them IIRC is the X format with either of the new parsers), but is this a correct assumption?
If I'm reading the DirectX specification correctly ( https://docs.microsoft.com/en-us/windows/win32/direct3d9/per-vertex-color-state ) then this overrides the material properties ambient etc. colors, and I would assume texturing still works in this case.

Comment (s520) [reply assets/Shaders/default.frag]: Oops. I fixed it.

Comment (s520) [reply assets/Shaders/default.frag]: wait a minute.
I need to fix it a little more.

**Comment by s520** [comment]
In older renderers, using two textures to specify an emission color disables lighting.
(It doesn't matter if the two textures are the same.)
This behavior is correct when the hack is enabled, but it is the same when the hack is disabled.
Lighting should be enabled if the hack is disabled.
Is this a known issue?

I looked at examples below in the ObjectViewer on the master branch.

### e.g. 1
```
LoadTexture, Daytime-base.png, Ref-night.png
SetEmissiveColor, 0, 255, 0
```
- Old renderer
DayNight_OR

- New renderer
DayNight_NR

### e.g. 2
```
LoadTexture, Daytime-base.png
SetEmissiveColor, 0, 255, 0
```
- Old renderer
Day_OR

- New renderer
Day_NR

### e.g. 3
```
SetColor, 0, 0, 255,
SetEmissiveColor, 0, 255, 0
```
- Old renderer
Color_OR

- New renderer
Color_NR

**Comment by leezer3** [comment]
I think you've probably found another one....

Logically, I'd expect the emissive color to be respected.

**Comment by s520** [comment]
I haven't been able to find where it's coming from.
Well, it seems better to think about it separately from this PR.

Therefore, there are no changes to this PR.

**Comment by leezer3** [comment]
Merged this into the emissive branch.

I can't see anything immediately broken, still testing before I merge into the main branch.
Have also added a command-line option to the viewers for good measure to make testing all of this a little easier:
```
/enablehacks
```

**Pull request merged**